### PR TITLE
Merge kde realinheritance

### PR DESCRIPTION
**Pull request by Plaristote**
These are the progresses I'm making merging the realinheritance branch.
I'm trying to split the major overhaul into smaller commits, while making sure not to break anything after each step.

Right now I have the first commit, and a small bit of the second one, which consisted in changing the prototype for createSimpleProperty, plus renaming that function to createProperty.

This is not especially up for a merge yet. I'm just opening this so that we can discuss the changes, and whether this merge is following the right direction.

### CODE REVIEW

File: src/qtcore/qml/elements/QtQuick/Row.js [hunk 1]
```diff
@@ -9,9 +9,9 @@ registerQmlType({
 function QMLRow(meta) {
     QMLPositioner.call(this, meta);
 
-    createProperty({ type: "enum", object: this, name: "layoutDirection" });
+    createProperty({ type: "enum", object: this, name: "layoutDirection", initialValue: 0 });
     this.layoutDirectionChanged.connect(this, this.layoutChildren);
-    this.layoutDirection = 0;
+    this.layoutDirectionChanged(0);
```
Comment (akreuzkamp) [inline src/qtcore/qml/elements/QtQuick/Row.js]: why do we need this line? Isn't this.layoutChildren(); rather what we want?

File: src/qtcore/qml/QMLPositioner.js [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
     QMLItem.call(this, meta);
 
-    createProperty({ type: "int", object: this, name: "spacing" });
+    createProperty({ type: "int", object: this, name: "spacing", initalValue: 0 });
     this.spacingChanged.connect(this, this.layoutChildren);
     this.childrenChanged.connect(this, this.layoutChildren);
     this.childrenChanged.connect(this, QMLPositioner.slotChildrenChanged);
 
-    this.spacing = 0;
+    this.spacingChanged(0);
```
Comment (akreuzkamp) [inline src/qtcore/qml/QMLPositioner.js]: why do we need this line? Isn't this.layoutChildren(); rather what we want?

**Comment by akreuzkamp** [comment]
looks (almost) good to me :)

File: src/qtcore/qml/elements/QtQuick/Row.js (hunk 1 above)
Comment (Plaristote) [inline src/qtcore/qml/elements/QtQuick/Row.js]: I'm trying to preserve the original behavior as best as I can.
At some places, the default value for the property were only set after connecting functions to the property's changed signal. In those cases, I've made sure the signals were still going to be triggered.
I'm not sure it does something in every cases : but I know that at least in some, it is important (for instance, I forgot it at some point on the Image item, and it broke an image test).

File: src/qtcore/qml/QMLPositioner.js (hunk 2 above)
Comment (Plaristote) [inline src/qtcore/qml/QMLPositioner.js]: Actually, forget my previous reaction : you're right, this.layoutChildren is more appropriate. I've fixed it.

**Comment by ChALkeR** [comment]
I'm not exactly sure about changing `createProperty` to receive all parameters inside an options object.

IMO it would be better to keep `type`/`object`/`name` arguments as they are now in master, and put optional arguments (`initalValue`) inside an options array, like this:
```js
createProperty("int", this, "spacing", { initalValue: 0 });
```

That would keep the code a little shorter and easier to write.

**Comment by ChALkeR** [comment]
About the `initialValue` for `string`/`list` — shouldn't that always be set to a valid string or list? If yes, shouldn't it default to `""` and `[]` inside the `createProperty` itself? I'm not about how Qt QML behaves yet.

**Comment by ChALkeR** [comment]
These commits LGTM:
 * «Fix some closure compiler warnings.»
 * «stores the current execution context in a variable»
 * «remove extra params from Qt.createQmlObject and Qt.createComponent»

**Comment by akreuzkamp** [comment]
Yes, properties could and should be default constructed automatically, nowadays. It's a matter of calling this.value = constructors[this.type](); inside the QMLProperty constructor (and maybe fixing some basic type constructors).

I'm open to either having createProperty arguments as one object or multiple arguments. :)

**Comment by ChALkeR** [comment]
@Plaristote Could you please split this into two pull requests, one with the `createProperty` changes (two commits) and the one with everything else, so we could discuss those separately? They should be independent, as it looks to me.

**Comment by Plaristote** [comment]
See #186 for that.

**Comment by ChALkeR** [comment]
I'm working on an alternate version of this.

**Comment by ChALkeR** [comment]
Update: see #200.

**Pull request closed without merging**